### PR TITLE
Card: migrate to TypeScript

### PR DESCRIPTION
**Pull request by torounit**
## What?

Part of https://github.com/WordPress/gutenberg/issues/35744

Converts the `Card` component to TypeScript.

## Testing Instructions

1. `npm run storybook:dev`
2. check `Card`

### CODE REVIEW

File: packages/components/src/card/card-header/component.tsx [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
-import { contextConnect } from '../../ui/context';
+import { contextConnect, WordPressComponentProps } from '../../ui/context';
 import { Flex } from '../../flex';
 import { useCardHeader } from './hook';
-
+import type { HeaderProps } from '../types';
 /**
- * @param {import('../../ui/context').WordPressComponentProps<import('../types').HeaderProps, 'div'>} props
- * @param {import('react').ForwardedRef<any>}                                                         forwardedRef
+ * External dependencies
```
Comment (ciampo) [inline packages/components/src/card/card-header/component.tsx]: External dependencies should be listed before internal dependencies

File: packages/components/src/card/card/hook.ts [hunk 2]
```diff
@@ -7,17 +7,20 @@ import { useMemo } from '@wordpress/element';
...
-	/**@type {import('../../ui/context').WordPressComponentProps<import('../types').Props, 'div'>} */
-	const propsToReturn = {
+type CardProps = WordPressComponentProps< Props, 'div' >;
+
+function useDeprecatedProps( {
+	elevation,
+	isElevated,
+	...otherProps
+}: CardProps ): CardProps {
+	const propsToReturn: CardProps = {
```
Comment (ciampo) [inline packages/components/src/card/card/hook.ts]: I believe the most correct way to code this hook is 

```suggestion
}: CardProps ) {
	const propsToReturn: Omit< CardProps, 'isElevated' > = {
```

1. Let's not add an explicit return value for the function, and instead rely on TypeScript inferring it
2. Let's be explicit about the fact that `isElevated` will be excluded from the return value of this hook

File: packages/components/CONTRIBUTING.md [hunk 3]
```diff
@@ -156,7 +156,7 @@ function Example(
 
 A couple of good examples of how hooks are used for composition are:
 
-- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.js);
+- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.ts);
```
Comment (ciampo) [inline packages/components/CONTRIBUTING.md:159]: It looks like there are more references left to `Card` files ending in `*.js` in this file, that also need updating

**Review by ciampo — COMMENTED** [review]
Hey @torounit , thank you for working on this!

On top of the inline comments, here's some additional feedback:

- Should we add a wrapper around the `Card` in the Storybook example?
- Currently, Storybook is not showing controls and documentation for the subcomponents — how can we fix that ? (cc @mirka )
- Can we also refactor the unit tests to TypeScript?

Thank you!

**Comment by mirka** [comment]
> * Currently, Storybook is not showing controls and documentation for the subcomponents — how can we fix that ? (cc @mirka )

Hmm, the subcomponents are showing for me. Maybe yours needed a cold restart?

https://user-images.githubusercontent.com/555336/183650847-343ca5a7-14ef-4761-9d4c-f6cf84f72339.mp4

(Subcomponents don't support controls, btw, so only the docs will show)

**Comment by ciampo** [comment]
Ah! It took me a while to find them!

https://user-images.githubusercontent.com/1083581/183933426-0d1d335b-4e37-4567-9dea-1f9b5553067b.mp4

> (Subcomponents don't support controls, btw, so only the docs will show)

That's unfortunate. Hopefully support will be added soon

**Comment by torounit** [comment]
@ciampo 

Sorry for the late reply, thanks for the review.

* Added Wrapper to the story.
* Fixed CONTRIBUTING.md.
* Refactored unit tests to TypeScript.

File: packages/components/CONTRIBUTING.md [hunk 4]
```diff
@@ -276,7 +276,7 @@ export function useCard( props ) {
 }
 
 //=========================================================================
-// Simplified snippet from `packages/components/src/card/card/component.js`
+// Simplified snippet from `packages/components/src/card/card/component.ts`
```
Comment (ciampo) [inline packages/components/CONTRIBUTING.md:279]: I found one more instance that needs renaming:

https://github.com/WordPress/gutenberg/blob/8f82fbc970a4250fee4dd7ac333aa890f69af16d/packages/components/CONTRIBUTING.md?plain=1#L316

File: packages/components/src/card/test/index.tsx [hunk 5]
```diff
@@ -56,9 +56,9 @@ describe( 'Card', () => {
 			const { container: containerSquared } = render(
 				<Card isRounded={ false }>Code is Poetry</Card>
 			);
-			expect( containerRounded.firstChild ).toMatchStyleDiffSnapshot(
-				containerSquared.firstChild
-			);
+			expect(
+				containerRounded.firstElementChild
+			).toMatchStyleDiffSnapshot( containerSquared.firstElementChild );
```
Comment (ciampo) [inline packages/components/src/card/test/index.tsx:61]: We also need to rename `packages/components/src/card/test/__snapshots__/index.js.snap` to `packages/components/src/card/test/__snapshots__/index.tsx.snap`

**Review by ciampo — COMMENTED** [review]
We're almost there! Just left a couple of minor comments.

Thank you for your patience!

**Comment by ciampo** [comment]
This PR may need a rebase to pass CI tests

**Comment by ciampo** [comment]
> This PR may need a rebase to pass CI tests

Ah, and a CHANGELOG entry (the "internal" section is ok)

**Review by ciampo — APPROVED** [review]

**Pull request merged**